File: path-finder-service/path_finder_service/feeders/redis_feeder.py
import asyncio
import json
from collections import defaultdict

import redis.asyncio as redis

from path_finder_service.jobs import JobRunner
from path_finder_service.storage.corpus import CorpusStore


class RedisStreamFeeder:
    def __init__(
        self,
        redis_url: str,
        stream_key: str,
        group_name: str,
        consumer_name: str,
        corpus_store: CorpusStore,
        job_runner: JobRunner,
        min_pages: int = 2,
    ):
        self._redis = redis.from_url(redis_url)
        self._stream_key = stream_key
        self._group_name = group_name
        self._consumer_name = consumer_name
        self._corpus = corpus_store
        self._runner = job_runner
        self._min_pages = max(min_pages, 2)
        self._counts: dict[str, int] = defaultdict(int)

    async def _ensure_group(self) -> None:
        try:
            await self._redis.xgroup_create(
                self._stream_key, self._group_name, id="0", mkstream=True
            )
        except redis.ResponseError as e:
            if "BUSYGROUP" not in str(e):
                raise

    async def start(self) -> None:
        await self._ensure_group()
        while True:
            messages = await self._redis.xreadgroup(
                self._group_name,
                self._consumer_name,
                {self._stream_key: ">"},
                count=10,
                block=5000,
            )
            for _, entries in messages:
                for msg_id, data in entries:
                    await self._process_message(data)
                    await self._redis.xack(
                        self._stream_key, self._group_name, msg_id
                    )

    async def _process_message(self, data: dict[bytes, bytes]) -> None:
        url = data[b"url"].decode("utf-8")
        html = data[b"html"].decode("utf-8")
        job_id = data[b"job_id"].decode("utf-8")

        index = self._counts[job_id]
        await self._corpus.put(job_id, index, url, html)
        self._counts[job_id] += 1

        if self._counts[job_id] >= self._min_pages:
            await self._runner.trigger(job_id)

    async def feed(self, url: str, html: str, job_id: str) -> None:
        await self._redis.xadd(
            self._stream_key,
            {"url": url, "html": html, "job_id": job_id},
        )

    async def force(self, job_id: str) -> None:
        await self._runner.trigger(job_id)

File: path-finder-service/internal/feeders/redis_feeder.go
package feeders

import (
	"context"
	"log/slog"
	"sync"

	"github.com/redis/go-redis/v9"

	"github.com/user/path-finder-service/internal/jobs"
	"github.com/user/path-finder-service/internal/storage"
)

type RedisStreamFeeder struct {
	rdb          *redis.Client
	streamKey    string
	groupName    string
	consumerName string
	corpus       *storage.CorpusStore
	runner       *jobs.JobRunner
	minPages     int
	mu           sync.Mutex
	counts       map[string]int
}

func NewRedisStreamFeeder(
	redisURL, streamKey, groupName, consumerName string,
	corpus *storage.CorpusStore,
	runner *jobs.JobRunner,
	minPages int,
) (*RedisStreamFeeder, error) {
	opts, err := redis.ParseURL(redisURL)
	if err != nil {
		return nil, err
	}
	if minPages < 2 {
		minPages = 2
	}
	return &RedisStreamFeeder{
		rdb:          redis.NewClient(opts),
		streamKey:    streamKey,
		groupName:    groupName,
		consumerName: consumerName,
		corpus:       corpus,
		runner:       runner,
		minPages:     minPages,
		counts:       make(map[string]int),
	}, nil
}

func (f *RedisStreamFeeder) Start(ctx context.Context) error {
	// Ensure consumer group
	_ = f.rdb.XGroupCreateMkStream(ctx, f.streamKey, f.groupName, "0").Err()

	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}

		streams, err := f.rdb.XReadGroup(ctx, &redis.XReadGroupArgs{
			Group:    f.groupName,
			Consumer: f.consumerName,
			Streams:  []string{f.streamKey, ">"},
			Count:    10,
			Block:    5000,
		}).Result()
		if err == redis.Nil {
			continue
		}
		if err != nil {
			slog.Error("xreadgroup error", "err", err)
			continue
		}

		for _, stream := range streams {
			for _, msg := range stream.Messages {
				f.processMessage(ctx, msg)
				f.rdb.XAck(ctx, f.streamKey, f.groupName, msg.ID)
			}
		}
	}
}

func (f *RedisStreamFeeder) processMessage(ctx context.Context, msg redis.XMessage) {
	url, _ := msg.Values["url"].(string)
	html, _ := msg.Values["html"].(string)
	jobID, _ := msg.Values["job_id"].(string)
	if jobID == "" {
		return
	}

	f.mu.Lock()
	index := f.counts[jobID]
	f.counts[jobID] = index + 1
	count := index + 1
	f.mu.Unlock()

	if err := f.corpus.Put(ctx, jobID, index, url, html); err != nil {
		slog.Error("corpus put failed", "job_id", jobID, "err", err)
		return
	}

	if count >= f.minPages {
		f.runner.Trigger(ctx, jobID)
	}
}

func (f *RedisStreamFeeder) Feed(ctx context.Context, url, html, jobID string) error {
	return f.rdb.XAdd(ctx, &redis.XAddArgs{
		Stream: f.streamKey,
		Values: map[string]interface{}{
			"url":    url,
			"html":   html,
			"job_id": jobID,
		},
	}).Err()
}

func (f *RedisStreamFeeder) Force(ctx context.Context, jobID string) {
	f.runner.Trigger(ctx, jobID)
}
